src/interface.jl

```diff
@@ -110,3 +110,19 @@ Check if `sys` has a constant structure. Constant structure systems do not chang
 number of variables or parameters over time.
 """
 constant_structure(sys) = constant_structure(symbolic_container(sys))
+
+"""
+    all_solvable_symbols(sys)
```

Comment:
This name isn't very informative. Also, I think we want it to be a singleton? `sol[allvariables]`, `sol[allsymbols]`? That lowers to these functions.

And for the API, the user needs to know what the symbol the value corresponds to, since they aren't giving it to you. I would think that this would need to return `Pair`s between `sym => value` since otherwise the ordering is simply undefined.

Reply:
Sure. I'll address this after fixing the RAT bugs in other repositories

Reply:
One thing to consider: `allvariables` will index variables and observed quantities, which I guess should be fine. `allsymbols` will also include parameters, which for e.g. solutions will return a `Vector{any}` since `sol[x]` is an array but `sol[p]` is a single value.

I'm also documenting the singletons as returning all variables/symbols *at the last timestep* so for future implementation of dynamic states the interface is clear.

Reply:
Or, wait, did you mean that `sol[allvariables]` should return a `Vector{Pair}`?

Reply:
Yes, I think so. I don't see any other way for it to be very useful. Or we should document how to easily get that pair.

Reply:
`sol[allvariables]` is a  time series, `sol[allvariables,1]`, `sol[allvariables,1:10]`, etc. should just act like the same thing as `sol[x,1]` just now with a generic symbolic.

Reply:
So `sol[allvariables]` is basically just `sol[variable_symbols(sol)]`? In which case, do we really need the singleton? What I imagined was `allvariables` is `variable_symbols` + the lhs of all observed equations. `allsymbols` is that + `parameter_symbols` and `independent_variable_symbols`. Complementary to the two singletons, we have `all_variable_symbols` which returns a vector of pairs from each variable symbols/observed symbol to its initial value and `all_symbols` which is the same but also maps parameter symbols/indepvar symbols to their value.

Reply:
We also can't require `all_variables(sol)` to return a vector of pairs, since e.g. `SymbolCache` doesn't store values.

Reply:
> What I imagined was allvariables is variable_symbols + the lhs of all observed equations.

Yes it's all of. Let's write out a whole thing:

- `solvedvariables`: Just the solved `u`
- `allvariables`: `u` + `observed`
- `allsymbols`: `u`+ `observed` + `p` + independent vars (good for making a table output!)

>  Complementary to the two singletons, we have all_variable_symbols which returns a vector of pairs from each variable symbols/observed symbol to its initial value and all_symbols which is the same but also maps parameter symbols/indepvar symbols to their value.

I think that's good.

> I don't quite understand the interface you're looking for here. sol[allvariables] is simple enough, if a bit redundant. Is the description allsymbols in line with what you thought of?

I just think it would be good to have a nice representation of `sol[:,i]` where `:` is this expanded "all variables" (since it's asked for often).

> We also can't require all_variables(sol) to return a vector of pairs, since e.g. SymbolCache doesn't store values.

What do you mean?

Reply:
You mentioned that `all_variables(sol)` needs to return a `Vector{Pair}` mapping each symbol to its (initial) value. However, that can't be a general implementation since we don't assume that everything implementing the interface has access to the values of symbols; it's a symbolic *indexing* interface, so it only is meant to provide indexes. `SymbolCache` is an example where it can't return a `Vector{Pair}` since it doesn't have access to the values of the variables.

I'd argue the values of the variables can be obtained using `getindex` on the types that implement it.

Reply:
I see, just make it return the values then. What's a way to get all of the symbols then as a vector generally? That would be a very crucial thing to have in order to know the ordering.

Reply:
It's the reverse, we can get all of the symbols but not their values.

Reply:
I see, but the singletons can be defined here for downstream to implement? Like how `Colon` works?

Reply:
Yeah

Reply:
I just realized: `allsymbols` can't include parameters because we're removing parameter indexing. i.e. if `sol[p]` isn't valid syntax then `sol[allsymbols]` can't be either.

Reply:
Got it. A separate `allparameters`?

Reply:
Yep. And no `allsymbols`? Or do we keep it for the independent variables?

Reply:
No `allsymbols`.

Reply:
No wait, we don't need `allparameters` either 😅 . We already have `parameter_values(sys)` which returns the parameter vector. `setp` support requires that the returned vector be a mutable reference to the actual parameter vector, so mutating it will mutate the parameters. `getp(sys, allparameters)(sol)` is just `parameter_values(sol)`. `setp(sys, allparameters)(sol, new_value)` is just `parameter_values(sol) .= new_value`

Reply:
Anything like this is something that should be described in tutorials. Make a small MTK model and show all of the tricks.